### PR TITLE
8234131: Miscellaneous changes imported from jsr166 CVS 2021-01

### PR DESCRIPTION
**Pull request by Martin-Buchholz**
/cc core-libs

---------
### Progress
- [x] Change must not contain extraneous whitespace
- [x] Commit message must refer to an issue
- [x] Change must be properly reviewed

### Issues
 * [JDK-8234131](https://bugs.openjdk.java.net/browse/JDK-8234131): Miscellaneous changes imported from jsr166 CVS 2021-01
 * [JDK-8257671](https://bugs.openjdk.java.net/browse/JDK-8257671): ThreadPoolExecutor.Discard*Policy: rejected tasks are not cancelled

### Reviewers
 * [Alan Bateman](https://openjdk.java.net/census#alanb) (@AlanBateman - **Reviewer**) ⚠️ Review applies to a96e284edc6d425aa3edf99a7ae40ff1cb829729
 * [Pavel Rappo](https://openjdk.java.net/census#prappo) (@pavelrappo - **Reviewer**) ⚠️ Review applies to a96e284edc6d425aa3edf99a7ae40ff1cb829729
 * [Doug Lea](https://openjdk.java.net/census#dl) (@DougLea - **Reviewer**)

### Download
`$ git fetch https://git.openjdk.java.net/jdk pull/1647/head:pull/1647`
`$ git checkout pull/1647`

### CODE REVIEW

**Comment by shipilev** [comment]
It would be nice to get it tested with GH Actions. "Checks" tabs should have those testing results automatically on push, but it is empty. Probably because your fork is not configured for it. Please go to https://github.com/Martin-Buchholz/jdk/actions -- and see if it says anything suspicious? Triggering the (only) workflow manually against your branch would help too.

File: src/java.base/share/classes/java/util/concurrent/ThreadPoolExecutor.java [hunk 1]
```diff
@@ -1544,11 +1551,13 @@ else if (delta > 0) {
             // As a heuristic, prestart enough new workers (up to new
             // core size) to handle the current number of tasks in
             // queue, but stop if queue becomes empty while doing so.
+            /*
             int k = Math.min(delta, workQueue.size());
             while (k-- > 0 && addWorker(null, true)) {
                 if (workQueue.isEmpty())
                     break;
             }
+            */
```
Comment (AlanBateman) [inline src/java.base/share/classes/java/util/concurrent/ThreadPoolExecutor.java:1560]: Is this meant to be commented out?

Comment (pavelrappo) [reply src/java.base/share/classes/java/util/concurrent/ThreadPoolExecutor.java:1560]: This code was also commented out in the original CVS repo; here's the diff: http://gee.cs.oswego.edu/cgi-bin/viewcvs.cgi/jsr166/src/main/java/util/concurrent/ThreadPoolExecutor.java?r1=1.193&r2=1.194

The message for the `1.194` revision suggests we should NOT expect code changes at all. I've double-checked my patch, which is at least partially responsible for the `1.194` revision, and couldn't find that commenting out part.

**Review by AlanBateman — APPROVED** [review]

File: src/java.base/share/classes/java/util/concurrent/ThreadPoolExecutor.java [hunk 2]
```diff
@@ -2081,7 +2088,18 @@ public void rejectedExecution(Runnable r, ThreadPoolExecutor e) {
...
+     * tasks to terminate, or failures must be recorded. Instead consider
+     * using a handler of the form:
+     * <pre> {@code
+     * new RejectedExecutionHandler() {
+     *   public void rejectedExecution(Runnable r, ThreadPoolExecutor e) {
+     *     Future<?> dropped = e.getQueue().poll();
+     *     if (dropped != null)
+     *        dropped.cancel(false); // also consider logging the failure
+     *     e.execute(r);             // retry
+     * }}}</pre>
```
Comment (pavelrappo) [inline src/java.base/share/classes/java/util/concurrent/ThreadPoolExecutor.java:2102]: I tried to reify that code snippet:

```
$ cat MySnippet.java 
import java.util.concurrent.Future;
import java.util.concurrent.RejectedExecutionHandler;
import java.util.concurrent.ThreadPoolExecutor;

public class MySnippet {
    public static void main(String[] args) {
        new RejectedExecutionHandler() {
            public void rejectedExecution(Runnable r, ThreadPoolExecutor e) {
                Future<?> dropped = e.getQueue().poll();
                if (dropped != null)
                    dropped.cancel(false); // also consider logging the failure
                e.execute(r);              // retry
            }
        };
    }
}
$ javac MySnippet.java 
MySnippet.java:9: error: incompatible types: Runnable cannot be converted to Future<?>
                Future<?> dropped = e.getQueue().poll();
                                                     ^
1 error
$ 
```
Separately, it seems that the `if` statement uses a 3-space indent which is surprising to see in the JSR 166 code base.

**Review by pavelrappo — APPROVED** [review]
The changes to doc comments look good.

Thanks for incorporating my earlier code snippet suggestions published on [concurrency-interest](http://cs.oswego.edu/pipermail/concurrency-interest/2020-November/017264.html)!

**Comment by JanecekPetr** [comment]
This also fixes issue [JDK-8257671](https://bugs.openjdk.java.net/browse/JDK-8257671): _ThreadPoolExecutor.Discard*Policy: rejected tasks are not cancelled_, by updating the `TPE.Discard*Policy` JavaDocs.

I filed it days before the code went in as I did not see the JavaDoc update in JDK, sorry for my impatience and thank you for the fix. I cannot update the Jira, please link it to this PR, too.

**Comment by Martin-Buchholz** [comment]
@JanecekPetr Thank you very much.  I've updated JIRA and this pull request for JDK-8257671

**Review by DougLea — APPROVED** [review]

**Pull request closed without merging**